src/combinators/mod.rs

```diff
@@ -1,115 +1,100 @@
 use gen::GenError;
-use std::io::{Cursor, Write};
+use std::io::Write;
 use std::fmt;
 
-pub trait SerializeFn<I>: Fn(I) -> Result<I, GenError> {}
+pub trait SerializeFn<I>: Fn(I) -> Result<(I, usize), GenError> {}
 
-impl<I, F:  Fn(I) ->Result<I, GenError>> SerializeFn<I> for F {}
+impl<I, F:  Fn(I) ->Result<(I, usize), GenError>> SerializeFn<I> for F {}
+
+pub trait Length {
+    fn length(&self) -> usize;
```

Comment:
Both these traits seem only be necessary for the `skip` combinator, and this here does not really give the length but how much can be skipped. Maybe that can be combined into a single `Skip` trait where the `skip` function returns a `Result` so that it can fail? Basically converting the `skip` combinator to a trait.

Reply:
Yeah, length should definitely be "AvailableLength" r something like that. When I added it I was wondering if it could have been useful elsewhere, but I probably can merge it in Skip

Reply:
I was thinking of changing `Skip` to be

```rust
pub trait Skip {
    fn skip(&mut self, s: usize) -> Result<(), GenError>;
}
```

or something similar that, so that it directly call it from `skip`. That would probably make the expected behaviour of the trait more clear as it would directly map to the `skip` combinator.